[SAH-70]: Implement User Onboarding

Maiz27 opened the pull request.

Maiz27 committed:
feat: reactor auth middleware

- add middleware to client app
- update next, react, react-dom, eslint versions to latest across apps

Maiz27 committed:
chore: refactor auth token api route

- add accessRules param to auth middleware func

Maiz27 committed:
feat: refactor multistep form hooks

- improve useCreateFormStore flexibility

mde3 reviewed (commented):
I see you've updated all the next version across the apps, will that have any impact

Maiz27 commented:
> I see you've updated all the next version across the apps, will that have any impact

@mde3 They might impact some minor parts of the code but it was needed since the different versions across all the apps and packages were causing unusual errors when interacting with each other

Maiz27 committed:
feat: add onboradingform controller

- improve redirect in middleware for new users
- add hasCompletedOnboarding to jwt callback
- remove redundant packages/lib/auth
- code refactors

Maiz27 committed:
feat: add Role Details Form

- add BusinessRoleDetails & SupplierRoleDetails forms
- add TextArea component
- improve nested error handling for Input, Select, Toggle, and Textarea components

Maiz27 committed:
feat: add Preference Form

- add BusinessPreference & SupplierPreference forms
- dynamic business types and supplier categories
- refactor Radio into RadioGroup
- refactor CheckBox into CheckBoxGroup

Maiz27 committed:
feat: add Summary Form

- refactor form constants to contants.ts

Maiz27 committed:
feat: onboard suppliers & businesses

Maiz27 committed:
feat: improve auth middleware redirects

- add loading state to fetched fields
- update sign-in URL callback

Maiz27 committed:
fix: agent app access rules build error

Emmanuel-Melon commented on packages/features/auth/forms/onboarding/Preferences.tsx:
```diff
@@ -0,0 +1,78 @@
+  };
+
+  const renderForm = () => {
+    switch (role) {
```
Using a switch statement is cool because we have a relatively small set of roles. We could probably refactor this to conditionally render content by setting the role in the Zustand store when we start adding more roles.

You could probably create an issue for this later.

Maiz27 replied:
Make sense, I used switch for the potential of new roles getting added. With it, adding new roles is much easier

Emmanuel-Melon commented on packages/features/auth/forms/onboarding/RoleDetails.tsx:
```diff
@@ -0,0 +1,61 @@
+    navigateToNextStep("preferences");
+  };
+
+  const renderForm = () => {
```
This is fine but we could consider reusing this function later on.

Emmanuel-Melon commented on packages/features/auth/forms/onboarding/Summary.tsx:
```diff
@@ -0,0 +1,277 @@
+  const role = formData.role as "supplier" | "business";
+  // Create schema in two steps to avoid TS error with
+  // computed properties (role) on combined schema.
+  const baseSchema = z.object({
```
This is neat!

Emmanuel-Melon commented on packages/features/auth/forms/onboarding/Summary.tsx:
```diff
@@ -0,0 +1,277 @@
+    if (!businessLoading && !supplierLoading) {
+      const { business, supplier, preference } = validatedInput;
+
+      switch (role) {
```
Third occurrence!

Emmanuel-Melon commented on packages/features/auth/forms/onboarding/constants.ts:
```diff
@@ -0,0 +1,95 @@
+  contactEmail: z.string().email("Invalid email"),
+  phoneNumber: z.string().refine(
+    (phone) => {
+      const internationalFormat = /^\+\d{1,3}\d{6,14}$/;
```
Let's extract this to a function because we need to use this in multiple places, i.e., mobile number verification, deposits/ withdrawals, and code/ password resets.

Emmanuel-Melon commented on packages/features/auth/lib/auth.ts:
```diff
@@ -61,12 +63,27 @@ const initNextAuth = (): AuthOptions => {
             "x-hasura-allowed-roles": ["user"],
             "x-hasura-default-role": "user",
-            "x-hasura-role": "user",
+            "x-hasura-role": role ?? "user",
```
Neat!

Emmanuel-Melon commented on packages/lib/hooks/formStores/useOnBoardingFormStore.tsx:
```diff
@@ -0,0 +1,50 @@
+  "summary",
+] as const;
+
+// eslint-disable-next-line react-hooks/rules-of-hooks
```
@Maiz27  we might have to do something about this later on. Should we create an issue to explore a possible refactor? 

I think this works fine but these warnings do exist for a reason and we might end up with implicit bugs. I can't think of any at the moment but it might be worth considering, what do you think?

Maiz27 replied:
Good idea!

Emmanuel-Melon commented on packages/ui/components/Radio.tsx:
```diff
@@ -1,36 +1,75 @@
-  name: string;
-  register: any;
-  errors: any;
+  icon?: any;
```
I think there's a type for icon

Emmanuel-Melon reviewed (approved):
Good work!

mde3 commented on packages/ui/components/Navbar.tsx:
Solid work my guy, I've wanted to work on improving the navbar component but just wanted to wait & view the changes you made then track back to improving it 

Dw it's nothing major just few tweaks so i guess that could go in a separate pr

Maiz27 replied:
No worries man, feel free to do it after this gets merged!
